kratos/containers/constraint_data.h

```diff
@@ -0,0 +1,352 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(CONSTRAINT_DATA_H)
+#define CONSTRAINT_DATA_H
+// System includes
+#include <vector>
+#include <unordered_set>
+#include <iostream>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include <boost/functional/hash.hpp>
+
+namespace Kratos
+{
+
+// ConstraintEquation class start
+/*
+*   This class stores the information regarding the constraint equation. Naming convenction is defined like this.
+*   (each object of this class will store one equation in the given form)
+*
+*   slaveDOF = w_1*masterDOF_1 + w_2*masterDOF_2 + ..... + w_n*masterDOF_n
+*   
+*   each slaveDOF and each of its masterDOF have the following attributes 
+*   a. dof ID
+*   b. dof KEY
+*   c. equation ID
+*
+*   one should be able to access the constraint equation both with (dofID , dofKEY) pair and/or equationID of the slave.
+*   This equation is imposed on the linear system of equations either element wise or on the global system.
+*
+*/
+class ConstraintEquation
+{
+  private:
+    typedef Dof<double> DofType;
+    struct MasterData
+    {
+        KRATOS_CLASS_POINTER_DEFINITION(MasterData);
+        MasterData(DofType const &rMasterDof, double Weight = 0.0) : mMasterWeight(Weight), mId(rMasterDof.Id()), mKey(rMasterDof.GetVariable().Key())
+        {
+        }
+        size_t MasterKey() { return mKey; }
+        double MasterWeight() { return mMasterWeight; }
+        void UpdateWeight(double iWeight) { mMasterWeight += iWeight; }
+        size_t MasterDofId() { return mId; }
+        void SetEquationId(size_t Id) { mEquationId = Id; }
+        size_t MasterEqId() { return mEquationId; }
+        void SetMasterEqId(size_t Id) { mEquationId = Id; }
+
+      private:
+        double mMasterWeight;
+        const size_t mId;
+        const size_t mKey;
+        size_t mEquationId;
+    };
+    typedef MasterData::Pointer MasterDataPointerType;
+    struct MasterHasher
+    {
+        size_t
+        operator()(const MasterDataPointerType &rObj) const
+        {
+            std::size_t seed = 0;
+            boost::hash_combine(seed, rObj->MasterDofId());
+            boost::hash_combine(seed, rObj->MasterKey());
+            return seed;
+        }
+    };
+
+    // Custom comparator that compares the string objects by length
+    struct MasterComparator
+    {
+        bool
+        operator()(const MasterDataPointerType &rObj1, const MasterDataPointerType &rObj2) const
+        {

```

Comment:
Because the unordered_set is of type ` MasterDataPointerType ` . So I cannot pass the object ref directly.  I wanted to make the set of MasterData objects directly but thought it will be a overhead as it has to be copied every time a new entry is added.